src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs

```diff
@@ -3,138 +3,700 @@
         /// </summary>
-        /// <param name="value">The parsed long value or default if failed.</param>
-        /// <returns>True if successfully parsed.</returns>
+        /// <param name="value">The parsed <see cref="ulong"/> value or default if failed.</param>
+        /// <param name="standardFormat">Expected format.</param>
+        /// <remarks>
+        /// <see cref="Utf8Parser.TryParse(ReadOnlySpan{byte}, out ulong, out int, char)"/> for supported formats.
+        /// </remarks>
+        /// <exception cref="FormatException">Thrown if the given format isn't supported.</exception>
+        /// <returns>Number of consumed bytes if successful, otherwise 0.</returns>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static unsafe bool TryParse(ref this BufferReader<byte> reader, out long value)
+        public static unsafe int TryParse(
+            ref this BufferReader<byte> reader,
+            out ulong value,
+            char standardFormat = '\0')
         {
             ReadOnlySpan<byte> unread = reader.UnreadSpan;
-            if (Utf8Parser.TryParse(unread, out value, out int consumed) && consumed < unread.Length)
+
+            if (!Utf8Parser.TryParse(unread, out value, out int consumed, standardFormat)
+                || consumed >= MaxParseBytes)
             {
+                // Failed to parse or we consumed more than we're willing to accept
+                return 0;
+            }
+            else if (consumed < unread.Length)
+            {
+                // The parser found a value it wouldn't consume so we have all useful data.
                 reader.Advance(consumed);
-                return true;
+                return consumed;
             }
 
-            return TryParseSlow(ref reader, out value);
+            // If we ate all of our unread there may be more valid digits.
+            // The typical max size for a long will be less than 27 bytes ("18,446,744,073,709,551,615").
+            return TryParseSlow(ref reader, out value, standardFormat, 27, UlongParser);
         }
 
-        private static unsafe bool TryParseSlow(ref BufferReader<byte> reader, out long value)
+        /// <summary>
+        /// Try to parse an <see cref="float"/> out of the reader (as UTF-8).
+        /// </summary>
+        /// <param name="value">The parsed <see cref="float"/> value or default if failed.</param>
+        /// <param name="standardFormat">Expected format.</param>
+        /// <remarks>
+        /// <see cref="Utf8Parser.TryParse(ReadOnlySpan{byte}, out float, out int, char)"/> for supported formats.
+        /// </remarks>
+        /// <exception cref="FormatException">Thrown if the given format isn't supported.</exception>
+        /// <returns>Number of consumed bytes if successful, otherwise 0.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe int TryParse(
```

Comment:
Why did we return number of consumed bytes as opposed to having an out parameter? I assume it's for perf, but it's not consistent with the general pattern that we already shipped. Is it really worth departing from the pattern?

Reply:
Yeah, but it wasn't a huge difference. I'm going to change it.